airflow/jobs/scheduler_job_runner.py

```diff
@@ -661,30 +675,52 @@ def _critical_section_enqueue_task_instances(self, session: Session) -> int:
         2. Change the state for the TIs above atomically.
         3. Enqueue the TIs in the executor.
 
-        HA note: This function is a "critical section" meaning that only a single executor process can execute
-        this function at the same time. This is achieved by doing ``SELECT ... from pool FOR UPDATE``. For DBs
-        that support NOWAIT, a "blocked" scheduler will skip this and continue on with other tasks (creating
-        new DAG runs, progressing TIs from None to SCHEDULED etc.); DBs that don't support this (such as
-        MariaDB or MySQL 5.x) the other schedulers will wait for the lock before continuing.
+        HA note: This function is a "critical section" meaning that only a single scheduler process can
+        execute this function at the same time. This is achieved by doing
+        ``SELECT ... from pool FOR UPDATE``. For DBs that support NOWAIT, a "blocked" scheduler will skip
+        this and continue on with other tasks (creating new DAG runs, progressing TIs from None to SCHEDULED
+        etc.); DBs that don't support this (such as MariaDB or MySQL 5.x) the other schedulers will wait for
+        the lock before continuing.
 
         :param session:
         :return: Number of task instance with state changed.
         """
+        # The user can either request a certain number of tis to schedule per main scheduler loop (default
+        # is non-zero). If that value has been set to zero, that means use the value of core.parallelism (or
+        # however many free slots are left). core.parallelism represents the max number of running TIs per
+        # scheduler. Historically this value was stored in the executor, who's job it was to control/enforce
+        # it. However, with multiple executors, any of which can run up to core.parallelism TIs individually,
+        # we need to make sure in the scheduler now that we don't schedule more than core.parallelism totally
+        # across all executors.
         if self.job.max_tis_per_query == 0:
-            max_tis = self.job.executor.slots_available
+            num_occupied_slots = sum([executor.slots_occupied for executor in self.job.executors])
+            max_tis = max(0, conf.getint("core", "parallelism") - num_occupied_slots)
         else:
-            max_tis = min(self.job.max_tis_per_query, self.job.executor.slots_available)
+            max_tis = self.job.max_tis_per_query
```

Comment:
Fair point, but I wonder if we actually want to ensure this later on when we're actually processing the result of the query. Since that gives us a greater selection of TIs to attempt to schedule from since many TIs end up not being schedulable due to various checks we have (pools, tis per dag, etc).
I can make your suggested change here and then we can observe how scheduling performs in real usecases and perhaps make this optimization later.

Reply:
Changes made and pushed to the PR!